enhancements/security/file-integrity-operator.md

```diff
@@ -0,0 +1,30 @@
+# Cluster Node File Integrity Operator
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs]
+
+## Open Questions [optional]
+
+## Summary
+
+This enhancement describes a new security feature for OpenShift. Many security-conscious customers want to be informed when files on a host's filesystem are modified in a way that is unexpected, as this may indicate an attack or compromise.  It proposes a "file-integrity-operator" that provides file integrity monitoring of select files on the host filesystems of the cluster nodes. It periodically runs a verification check on the watched files and provides logs of any changes.
```

Comment:
Summarizing some points I raised in internal discussion: I just don't see this implementation as worth it.

Security is full of tradeoffs - I can add three locks to my front door for example.  Is it worth it?  Probably not...it'd be a daily usability hit for a very marginal benefit.

Humans (including humans attempting to compromise computer systems) will follow the path of least resistance.  If they see 3 locks on my front door, they'll check the back door, where I didn't put 3 locks.

In this case, the back door is basically killing or compromising the AIDE daemonset.

This is going to be a very minor speedup to any attacker that has studied the system beforehand.

Further, it raises the risk of a lot of false positives.  How for example would an AIDE system distinguish between "attacker compromised `/usr/bin/bash`" and "OSTree update changed `/usr/bin/bash`".  Similarly for files in `/etc` that may be changed via MachineConfig, or certificates on the host that end up being rotated, etc.

Really a lot of this boils down to:

> providing a log of files that have been modified since the initial run of the DaemonSet pods.

Is just not even close to sufficient.

Finally, another thing is that "periodically scan the whole filesystem" is a known way to cause performance hits.  It means that files that were unused are suddenly brought into the page cache, potentially evicting hot files.  It causes I/O contention.

I understand we're trying to meet a compliance standard, and we also can't let the perfect be the enemy of the good.  We have to start somewhere, and I (and our customers) certainly appreciate the efforts here.

But my bottom line is that this implementation overall will cause more problems (in false positives and also periodic perf hits) than it will solve.

Reply:
> Summarizing some points I raised in internal discussion: I just don't see this implementation as worth it.
> 
> Security is full of tradeoffs - I can add three locks to my front door for example. Is it worth it? Probably not...it'd be a daily usability hit for a very marginal benefit.
> 
> Humans (including humans attempting to compromise computer systems) will follow the path of least resistance. If they see 3 locks on my front door, they'll check the back door, where I didn't put 3 locks.
> 
> In this case, the back door is basically killing or compromising the AIDE daemonset.
> 
> This is going to be a very minor speedup to any attacker that has studied the system beforehand.
> 

While this might be true, the same could be said about basically any OS-level security component that is not backed by a HSM or similar hardware root of trust, and even then that is just moving the goal post - you have to then trust that your hardware vendor is not selling its keys to a nation state. Given a sophisticated enough attacker it all falls apart, so you can't really include this worst case scenario when developing a threat model. 

Like Juan mentioned there is the possibility of extending this to different file integrity providers but we want to tackle the FedRAMP "moderate" baseline initially (we also would not want to make the HSM/TPM a barrier for entry to be compliant if the spec does not call for it). Even if at a practical level the AIDE solution only provides the ability to better post-mortem a compromise because you have a log of the files that changed, that is still valuable to the organization.

> Further, it raises the risk of a lot of false positives. How for example would an AIDE system distinguish between "attacker compromised `/usr/bin/bash`" and "OSTree update changed `/usr/bin/bash`". Similarly for files in `/etc` that may be changed via MachineConfig, or certificates on the host that end up being rotated, etc.
>  

We will need to come up with a good strategy for false positives. In the case of an OSTree update, this sounds like something the file integrity operator might be able to handle if it can detect that there was a cluster update and update the checksum database.